scipy/optimize/optimize.py

```diff
@@ -2417,17 +2419,105 @@ def bracket(func, xa=0.0, xb=1.0, args=(), grow_limit=110.0, maxiter=1000):
+    lower_bound, upper_bound = lower_bound[nonzero], upper_bound[nonzero]
+    x0, alpha = x0[nonzero], alpha[nonzero]
+    low = (lower_bound - x0) / alpha
+    high = (upper_bound - x0) / alpha
+
+    # positive and negative indices
+    pos, neg = alpha > 0, alpha < 0
+
+    # inf * 0 = nan, but gives a RuntimeWarning when in array multiplication.
+    # In our case, we want inf * 0 to be 0.
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=RuntimeWarning)
+
+        lmin_pos = low * pos
+        lmin_neg = high * neg
+        lmax_pos = high * pos
+        lmax_neg = low * neg
+
+    lmin_pos[np.isnan(lmin_pos)] = 0
+    lmin_neg[np.isnan(lmin_neg)] = 0
+    lmin = np.max(lmin_pos + lmin_neg)
+    lmax_pos[np.isnan(lmax_pos)] = 0
+    lmax_neg[np.isnan(lmax_neg)] = 0
+    lmax = np.min(lmax_pos + lmax_neg)
+
+    # if x0 is outside the bounds, then it is possible that there is 
+    # no way to get back in the bounds for the parameters being updated
+    # with the current direction alpha.
+    # when this happens, lmax >= lmin.
+    # If this is the case, then we can just return (0, 0)
+    return (lmin, lmax) if lmax >= lmin else (0, 0)
+
+
+def _linesearch_powell(func, p, xi, lower_bound=None, upper_bound=None,
+                       fval=None, tol=1e-3):
     """Line-search algorithm using fminbound.
 
-    Find the minimium of the function ``func(x0+ alpha*direc)``.
+    Find the minimium of the function ``func(x0 + alpha*direc)``.
+    
+    fval = func(p), might as well avoid recomputing it (limit fevals).
+
+    lower_bound and upper_bound should be np.arrays, where
+    lower_bound[i] is the lower bounds on the i^th parameter,
+    etc.
 
     """
     def myfunc(alpha):
         return func(p + alpha*xi)
-    alpha_min, fret, iter, num = brent(myfunc, full_output=1, tol=tol)
-    xi = alpha_min*xi
-    return squeeze(fret), p + xi, xi
+
+    # if xi is zero, then don't optimize
+    if not np.any(xi):
+        return ((fval, p, xi) if fval is not None else (func(p), p, xi))
+    elif lower_bound is None and upper_bound is None:
+        alpha_min, fret, _, _ = brent(myfunc, full_output=1, tol=tol)
+        xi = alpha_min * xi
+        return squeeze(fret), p + xi, xi
+    else:
+        bound = _line_for_search(p, xi, lower_bound, upper_bound)
+        if bound[0] == -np.inf and bound[1] == np.inf:
+            # equivalent to unbounded
+            return _linesearch_powell(func, p, xi, fval=fval, tol=tol)
+        elif bound[0] != -np.inf and bound[1] != np.inf:
+            # we can use a bounded scalar minimization
+            res = _minimize_scalar_bounded(myfunc, bound, xatol=tol / 100)
+            xi = res.x * xi
+            return squeeze(res.fun), p + xi, xi
+        else:
+            # only bounded on one side. use the tangent function to convert
+            # the infinity bound to a finite bound.
+            bound = np.arctan(bound[0]), np.arctan(bound[1])
```

Comment:
Do you have a reference for the `np.arctan` scaling and it's accuracy/stability? As bounds get much larger in magnitude `np.arctan(bound)` won't change by all that much, I'm guessing this will degrade the ability of a minimizer to work because it's like doing finite differences with float16/32 instead of float64. Is this a regular transform for optimisation problems?

 I'm aware of http://cars9.uchicago.edu/software/python/lmfit_MinimizerResult/bounds.html for single side bounds, but haven't seen this used yet.

Reply:
I do not have a reference for this, however I don't immediately see why it would be worse than the ``sin`` transformation used in the link you referenced. Do you know why it would be? I don't know what is the "standard method", however I suspect that any transformation from bounded-on-one-side to bounded-on-both-sides will have numerical instabilities. Perhaps it should be noted in the documentation somewhere that one may encounter numerical instabilities for certain types of ranges?